### DISCUSSION TITLE
Is it possible to use OkHttp cache with Coil 2?

### DISCUSSION THREAD

**Discussion by gmk57** [coil-kt/coil#1225] [Q&A]
We're migrating from Picasso to Coil for an app with a chat-like screen. We need to show images on the screen as well as download files (including images) for passing them to other apps, minimizing network and storage usage.

Files are downloaded through OkHttp, with disk cache configured.
For Picasso we've enabled integration with the same OkHttp instance (`OkHttp3Downloader(okHttpClient)`), which results in downloading and storing images only once for all these tasks.

AFAIK, the same was possible with Coil 1.x, but not with Coil 2. The docs specifically says: "*You **should not** use OkHttp's `Cache` with Coil 2.0 as the cache can be corrupted if a thread is interrupted while writing to it*"

That sounds scary. Should we opt for Coil 1.4 or does that issue affect 1.x too? Are there any better ways to achieve our goals?

**Comment by colinrtwhite** [comment]
The main reason Coil uses its own disk cache in 2.0 is because Coil uses thread interruption while decoding to cancel much faster. Part of reason Picasso and Coil 1.x are able to use OkHttp's disk cache is because they don't use thread interruption (which is slower). Additionally, OkHttp's cache doesn't provide direct file access which is sometimes needed.

If you use OkHttp's `Cache` with Coil 2.0 it's possible to interrupt the thread while writing to the disk journal, which only temporarily corrupts the disk journal. All reads to the disk cache will return `null` while it's temporarily corrupted. It'll be rebuilt in a subsequent app launch (when the disk cache is re-initialized).

Overall, I'd recommend only using Coil's custom `DiskCache` with Coil 2.0. Coil 1.4 will still work, but it's missing a lot of benefits in 2.0.

**Reply by gmk57** [reply]
Thanks for your detailed response! I was looking for this "motivation" part in Coil 2.0 docs, but haven't found it.

It's good news that cache will be rebuilt eventually, but working without cache until app restart probably won't be the best UX.

So in Coil 2 this thread interruption mechanism is not configurable, correct?

Then maybe it's possible to use both OkHttp's cache and Coil's cache? This way we'll store file twice, but download it only once.

The most efficient solution would probably be to use only Coils's cache for images, but eventually we need a local `Uri` for external apps (`Intent.ACTION_VIEW`). We could [preload image to disk](https://coil-kt.github.io/coil/getting_started/#preloading), then [get file path from `DiskCache`](https://github.com/coil-kt/coil/issues/100#issuecomment-939691688) and convert it to `Uri` with `FileProvider`, does that sound good?

**Reply by gmk57** [reply]
And one more question: we use a shared OkHttp client (for reusing connection pool, thread pools, interceptors, etc), and it needs a disk cache for non-image files. Should we then configure Coil 2.x along these lines:
`.okHttpClient { client.newBuilder().cache(null).build() }` ?
